Fix crash in Document::topmostAutoPopover()

chirags27 opened the pull request:
#### db12f6fd8603d4a56bc07e7072aae0c934fa1f15
<pre>
Fix crash in Document::topmostAutoPopover()
<a href="https://bugs.webkit.org/show_bug.cgi?id=253597">https://bugs.webkit.org/show_bug.cgi?id=253597</a>
rdar://106423630

Reviewed by Tim Nguyen and Chris Dumez.

This change fixes a crash introduced recently by making sure candidate
is non-null before dereferencing it.

* LayoutTests/fast/html/popover-with-fullscreen-crash-expected.txt: Added.
* LayoutTests/fast/html/popover-with-fullscreen-crash.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::topmostAutoPopover const):

Canonical link: <a href="https://commits.webkit.org/261391@main">https://commits.webkit.org/261391@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31d9b1e75ba84e2e6f0bcf6bb1a539e7b0ce515c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111564 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20697 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/169 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3326 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120324 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115623 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22054 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11784 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117327 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/99534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45162 "layout-tests (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13186 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/89/builds/109 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86981 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13681 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19126 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52094 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7931 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15662 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/html/popover-with-fullscreen-crash.html:
```diff
@@ -0,0 +1,17 @@
+      testRunner.notifyDone();
+  };
+</script>
+<div id="d" popover=""></div>
```
```suggestion
    const svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
    document.body.append(svg);
    await svg.requestFullscreen();
    document.querySelector("[popover]").showPopover();
    document.body.innerHTML = 'PASS if no crash';
    if (window.testRunner)
      testRunner.notifyDone();
  };
</script>
<div popover="auto"></div>
```

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8955,7 +8955,7 @@ HTMLDialogElement* Document::activeModalDialog() const
 {
     for (auto& element : makeReversedRange(m_topLayerElements)) {
-        if (auto* candidate = dynamicDowncast<HTMLElement>(element.get()); candidate->popoverState() == PopoverState::Auto)
+        if (auto* candidate = dynamicDowncast<HTMLElement>(element.get()); candidate && (candidate->popoverState() == PopoverState::Auto))
```
I don't think you need the extra inner brackets?

cdumez reviewed (approved):
LGTM

chirags27 committed:
Fix crash in Document::topmostAutoPopover()
https://bugs.webkit.org/show_bug.cgi?id=253597
rdar://106423630

Reviewed by Tim Nguyen and Chris Dumez.

This change fixes a crash introduced recently by making sure candidate
is non-null before dereferencing it.

* LayoutTests/fast/html/popover-with-fullscreen-crash-expected.txt: Added.
* LayoutTests/fast/html/popover-with-fullscreen-crash.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::topmostAutoPopover const):

Canonical link: https://commits.webkit.org/261391@main
